core/src/test/scala/org/apache/spark/sql/delta/optimize/OptimizeCompactionSuite.scala

```diff
@@ -311,6 +312,42 @@ trait OptimizeCompactionSuiteBase extends QueryTest
     }
   }
 
+  test("optimize - abort before all tasks complete") {
+    val numPartitions = 2
+    withTempDir { tempDir =>
+      spark.range(100)
+          .withColumn("pCol", 'id % numPartitions)
+          .repartition(10)
+          .write
+          .format("delta")
+          .partitionBy("pCol")
+          .save(tempDir.getAbsolutePath)
+
+      // Delete one of the partitions manually to force an exception.
+      // This relies on the partition 0 being compacted first, if it is not
+      // this will hang.
+      val partition = new File(tempDir, "pCol=0")
```

Comment:
This test only works if `pCol=0` is the first partition submitted to be compacted, because if it's not, the future for compacting `pCol=1` will just block waiting for another thread to try to start writing. This seems kinda non-deterministic since there's no actual sorting of the partitions before compacting, so not sure if there's a way to improve this or if it's deterministic enough

Reply:
Can you check if overriding the `BlockWritesAbstractFileSystem` to introduce failures based on partition id or path is more deterministic?

Reply:
Hi, just to clarify what want to test here:
1. We don't start any new jobs if we encounter any failure(s) in the currently running jobs
2. Any currently running jobs will be aborted as soon as we encounter a failure in one of the currently running jobs (not sure if the ThreadUtils utility method handles that)

Wondering if this testcase tests any of the above?

Here is the test I think that will verify (1) above is working
- Start optimize job that has more than the `DELTA_OPTIMIZE_MAX_THREADS` (example: 5 jobs, max_threads 2)
- Let the optimize complete 2 jobs.
- When running the third job, fail it (By creating a new FileSystem that overrides `BlockWritesAbstractFileSystem` and injects failure when trying to create a file or read a file - you look at the file path to decide ex. pCol=3. Also in this FileSystem implementation, you can track how many jobs are started to verify that we are not starting any new jobs once we encounter a failure.

Let me know if this is not clear.

Reply:
The existing test should effectively show that already. I tell the BlockWrites to only write once two are active, and the fact that the exception is thrown shows that we have continued passed the parmap call (which under the hood kills the thread pool so there can't be any more jobs running).

Would adding a third partition, and then checking that the third partition never starts (by `blockLatch.getCount() == 1`) suffice? In your example it has the same potential for non-deterministic behavior if it's not guaranteed what the ordering of partitioning is.

Reply:
Tried that and seems to be non-deterministic, I'll see if I can do something where it fails after the nth `create` call

Reply:
And to clarify, new jobs _will_ be started after a failure until the failure's future is resolved (i.e. all tasks submitted before it have finished), and then at that point the thread pool will be killed and everything stopped. Only alternative is trying to do a polling based mechanism to resolves futures in the order they finish